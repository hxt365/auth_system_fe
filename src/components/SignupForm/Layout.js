const formItemLayout = {
  labelCol: {
    sm: { span: 5 },
  },
  wrapperCol: {
    sm: { span: 14 },
  },
};

const tailFormItemLayout = {
  wrapperCol: {
    sm: {
      span: 7,
      offset: 8,
    },
  },
};

const firstNameItemLayout = {
  labelCol: {
    sm: { span: 10 },
  },
  wrapperCol: {
    sm: { span: 13 },
  },
};

const lastNameItemLayout = {
  wrapperCol: {
    sm: {
      span: 14,
    },
  },
};

export { firstNameItemLayout, lastNameItemLayout, tailFormItemLayout, formItemLayout };
