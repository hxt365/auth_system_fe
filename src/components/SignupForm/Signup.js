import React from 'react';

function SignupForm() {
  return <p>SignupForm</p>;
}

export default SignupForm;
