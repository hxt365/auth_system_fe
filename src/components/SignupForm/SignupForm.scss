.signup-form {
  max-width: 70rem;
  margin: auto;

  .row .ant-form-item {
    flex: 1;

    &:last-child {
      flex: 1;
    }
  }
}
