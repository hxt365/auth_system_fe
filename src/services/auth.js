// @flow

import {
  LOGIN,
  LOGOUT,
  REFRESH,
  RESET_PASSWORD,
  CHANGE_PASSWORD,
  SIGNUP,
  CONFIRM_EMAIL,
} from 'constants/api';

import type {
  loginFormType,
  resetPasswordType,
  changePasswordFormType,
  signupFormType,
} from 'type';

import customAxios from './axios';

const login = async (data: loginFormType) => {
  const res = await customAxios.post(LOGIN, data).catch(err => err);
  return res;
};

const logout = async () => {
  const res = await customAxios.post(LOGOUT).catch(err => err);
  return res;
};

const resetPassword = async (data: resetPasswordType) => {
  const res = await customAxios.post(RESET_PASSWORD, data).catch(err => err);
  return res;
};

const refresh = async () => {
  const res = await customAxios.post(REFRESH).catch(err => err);
  return res;
};

const changePassword = async (data: changePasswordFormType) => {
  const res = await customAxios.post(CHANGE_PASSWORD, data).catch(err => err);
  return res;
};

const signup = async (data: signupFormType) => {
  const res = await customAxios.post(SIGNUP, data).catch(err => err);
  return res;
};

const confirmEmail = async (token: String) => {
  const link = `${CONFIRM_EMAIL + token}/`;
  const res = await customAxios.post(link).catch(err => err);
  return res;
};

export default {
  login,
  logout,
  refresh,
  resetPassword,
  changePassword,
  signup,
  confirmEmail,
};
