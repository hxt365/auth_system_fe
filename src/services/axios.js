// @flow

import axios from 'axios';

export const BASE_API_URL: String = `${process.env.REACT_APP_BASE_API_URL}`;

const customAxios = axios.create({
  baseURL: BASE_API_URL,
  withCredentials: true,
  xsrfCookieName: 'csrftoken',
  xsrfHeaderName: 'X-CSRFToken',
});

customAxios.interceptors.response.use(
  response => response,
  error => {
    // Add logic to handle error here
    console.log(error.response);
    return Promise.reject(error);
  },
);

export default customAxios;
