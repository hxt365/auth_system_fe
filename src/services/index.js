export { default as authServices } from './auth';
