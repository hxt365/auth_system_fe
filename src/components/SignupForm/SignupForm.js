// @flow

import React, { useState, useContext } from 'react';
import { Form, Input, Tooltip, Icon, Checkbox, Button, Row } from 'antd';
import './SignupForm.scss';
import { AppContext } from 'components/AppLayout';
import { SUCCESS, ERROR } from 'constants/message';
import { authServices } from 'services';
import { LOGIN } from 'constants/route';
import { minLength, maxLength, containsValidCharacters, onlyLetters } from 'validators';
import { signupFormType } from 'type';
import {
  firstNameItemLayout,
  lastNameItemLayout,
  tailFormItemLayout,
  formItemLayout,
} from './Layout';

type PropsType = {
  form: any,
};

function SignupForm(props: PropsType) {
  const state = useContext(AppContext);

  const [dirty, setDirty] = useState(false);

  const { form } = props;
  const { getFieldDecorator } = form;

  const signup = async (data: signupFormType) => {
    const res = await authServices.signup(data);
    if (res.status === 201) {
      state.setMessage({
        status: SUCCESS,
        title:
          // eslint-disable-next-line max-len
          'You successfully registered! We sent you an verification email, just check it out nowwwww',
        button: 'Thank you, admin!',
        redirect: LOGIN,
      });
    } else {
      state.setMessage({
        status: ERROR,
        title: 'Username or email was taken',
        button: 'Let me try again',
        redirect: '',
      });
    }
  };

  const handleSubmit = e => {
    e.preventDefault();
    props.form.validateFields((err, values) => {
      if (!err)
        signup({
          username: values.username,
          firstName: values.firstName,
          lastName: values.lastName,
          email: values.email,
          password: values.password,
          password_2: values.confirm,
        });
    });
  };

  const handleConfirmBlur = e => {
    const { value } = e.target;
    setDirty(dirty || !!value);
  };

  const compareToFirstPassword = (rule, value, callback) => {
    if (value && value !== form.getFieldValue('password')) {
      callback('Two passwords that you enter are inconsistent!');
    } else {
      callback();
    }
  };

  const validateToNextPassword = (rule, value, callback) => {
    if (value && dirty) {
      form.validateFields(['confirm'], { force: true });
    }
    callback();
  };

  const validateUsername = (rule, value, callback) => {
    if (value) {
      if (!containsValidCharacters(value))
        callback('Username may contain only letters, numbers, and @/./+/-/_ characters');
      if (!minLength(value, 6)) callback('Ensure username has at least 6 characters');
      if (!maxLength(value, 150)) callback('Ensure username has no more than 150 characters');
    }
    callback();
  };

  const validatePassword = (rule, value, callback) => {
    if (value) {
      if (!minLength(value, 6)) callback('Ensure password has at least 6 characters');
      if (!maxLength(value, 128)) callback('Ensure password has no more than 128 characters');
    }
    callback();
  };

  const validateName = (rule, value, callback) => {
    if (value) {
      if (!onlyLetters(value)) callback('Invalid name');
      if (!maxLength(value, 15))
        callback('Ensure first nam/ last name has no more than 15 characters');
    }
    callback();
  };

  return (
    <Form {...formItemLayout} onSubmit={handleSubmit} className="signup-form">
      <Row type="flex" className="row">
        <Form.Item label="Name" {...firstNameItemLayout}>
          {getFieldDecorator('firstName', {
            rules: [
              { required: true, message: 'Please input your first name!', whitespace: true },
              { validator: validateName },
            ],
          })(<Input placeholder="First name" />)}
        </Form.Item>
        <Form.Item {...lastNameItemLayout}>
          {getFieldDecorator('lastName', {
            rules: [
              { required: true, message: 'Please input your last name!', whitespace: true },
              { validator: validateName },
            ],
          })(<Input placeholder="Last name" />)}
        </Form.Item>
      </Row>
      <Form.Item
        label={
          // eslint-disable-next-line react/jsx-wrap-multilines
          <span>
            Username&nbsp;
            <Tooltip title="What do you want others to call you?">
              <Icon type="question-circle-o" />
            </Tooltip>
          </span>
        }
        hasFeedback
      >
        {getFieldDecorator('username', {
          rules: [
            { required: true, message: 'Please input your username!' },
            { validator: validateUsername },
          ],
        })(<Input />)}
      </Form.Item>
      <Form.Item label="E-mail">
        {getFieldDecorator('email', {
          rules: [
            {
              type: 'email',
              message: 'The input is not valid E-mail!',
            },
            {
              required: true,
              message: 'Please input your E-mail!',
            },
          ],
        })(<Input />)}
      </Form.Item>
      <Form.Item label="Password" hasFeedback>
        {getFieldDecorator('password', {
          rules: [
            { required: true, message: 'Please input your password!' },
            {
              validator: validateToNextPassword,
            },
            {
              validator: validatePassword,
            },
          ],
        })(<Input.Password />)}
      </Form.Item>
      <Form.Item label="Confirm Password" hasFeedback>
        {getFieldDecorator('confirm', {
          rules: [
            {
              required: true,
              message: 'Please confirm your password!',
            },
            {
              validator: compareToFirstPassword,
            },
          ],
        })(<Input.Password onBlur={handleConfirmBlur} />)}
      </Form.Item>
      <Form.Item {...tailFormItemLayout}>
        {getFieldDecorator('agreement', {
          valuePropName: 'checked',
        })(
          <Checkbox>
            I have read the&nbsp;
            <a href="https://teko.vn/home/" target="_blank" rel="noopener noreferrer">
              agreement
            </a>
          </Checkbox>,
        )}
      </Form.Item>
      <Form.Item {...tailFormItemLayout}>
        <Button
          type="primary"
          onClick={handleSubmit}
          id="signup-form__button"
          disabled={!form.getFieldValue('agreement')}
        >
          Register
        </Button>
      </Form.Item>
    </Form>
  );
}

const wrapper = Form.create({ name: 'signupForm' })(SignupForm);

export default wrapper;
