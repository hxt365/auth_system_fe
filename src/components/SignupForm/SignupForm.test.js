import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { authServices } from 'services';
import SignupForm from './SignupForm';

describe('test SignupForm', () => {
  it('test snapshot', () => {
    const wrapper = shallow(<SignupForm />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // eslint-disable-next-line max-len
  it('submithandler called only once with username, firstname, lastname, email, password and confirm', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup').mockImplementation(() => {});
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    expect(usernameInput).toHaveLength(1);
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    expect(firstNameInput).toHaveLength(1);
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    expect(lastNameInput).toHaveLength(1);
    const emailInput = wrapper.find('input#signupForm_email');
    expect(emailInput).toHaveLength(1);
    const passwordInput = wrapper.find('input#signupForm_password');
    expect(passwordInput).toHaveLength(1);
    const confirmInput = wrapper.find('input#signupForm_confirm');
    expect(confirmInput).toHaveLength(1);
    const agreement = wrapper.find('input#signupForm_agreement');
    expect(agreement).toHaveLength(1);
    const username = 'username';
    const firstName = 'Testing';
    const lastName = 'Testing';
    const email = 'testing@example.com';
    const password = 'password';
    const confirm = 'password';
    usernameInput.simulate('change', { target: { value: username } });
    firstNameInput.simulate('change', { target: { value: firstName } });
    lastNameInput.simulate('change', { target: { value: lastName } });
    emailInput.simulate('change', { target: { value: email } });
    passwordInput.simulate('change', { target: { value: password } });
    confirmInput.simulate('change', { target: { value: confirm } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(1);
    expect(signupFormSpy).toHaveBeenCalledWith({
      username,
      firstName,
      lastName,
      email,
      password,
      password_2: confirm,
    });
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of disagreement', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of empty first name', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: '' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of first name having more than 15 characters', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: 'a'.repeat(16) } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of empty last name having more than 15 characters', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'a'.repeat(16) } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of empty last name', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: '' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of empty last name', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: '' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of invalid first name', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: '123@sth_' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of invalid last name', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'username' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: '123@sth_' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of empty username', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: '' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of username having less than 6 characters', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: '12345' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of username having more than 150 characters', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'a'.repeat(151) } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of invalid username', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'hello&gudbye' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'password' } });
    confirmInput.simulate('change', { target: { value: 'password' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of password having less than 6 characters', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'testing' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: '12345' } });
    confirmInput.simulate('change', { target: { value: '12345' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of password having more than 128 characters', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'testing' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: '1'.repeat(129) } });
    confirmInput.simulate('change', { target: { value: '1'.repeat(129) } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });

  it('submithandler not called because of unconsistent passwords', () => {
    const signupFormSpy = jest.spyOn(authServices, 'signup');
    const wrapper = mount(
      <BrowserRouter>
        <SignupForm />
      </BrowserRouter>,
    );
    const usernameInput = wrapper.find('input#signupForm_username');
    const firstNameInput = wrapper.find('input#signupForm_firstName');
    const lastNameInput = wrapper.find('input#signupForm_lastName');
    const emailInput = wrapper.find('input#signupForm_email');
    const passwordInput = wrapper.find('input#signupForm_password');
    const confirmInput = wrapper.find('input#signupForm_confirm');
    const agreement = wrapper.find('input#signupForm_agreement');
    usernameInput.simulate('change', { target: { value: 'testing' } });
    firstNameInput.simulate('change', { target: { value: 'Testing' } });
    lastNameInput.simulate('change', { target: { value: 'Testing' } });
    emailInput.simulate('change', { target: { value: 'testing@example.com' } });
    passwordInput.simulate('change', { target: { value: 'rightpassword' } });
    confirmInput.simulate('change', { target: { value: 'wrongpassword' } });
    agreement.simulate('change', { target: { checked: true } });
    const btn = wrapper.find('button#signup-form__button');
    expect(btn).toHaveLength(1);
    btn.simulate('click');
    expect(signupFormSpy).toHaveBeenCalledTimes(0);
    signupFormSpy.mockClear();
    wrapper.unmount();
  });
});
