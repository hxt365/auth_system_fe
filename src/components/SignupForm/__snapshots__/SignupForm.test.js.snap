// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SignupForm test snapshot 1`] = `
<SignupForm
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
/>
`;
